File: api/src/main/kotlin/org/team619/rutor/converter/MainPagePlainConverter.java
package org.team619.rutor.converter;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.team619.rutor.core.Converter;
import org.team619.rutor.core.Logger;
import org.team619.rutor.model.MainPlainPage;
import org.team619.rutor.model.Row;

import java.io.Serializable;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Created by BORIS on 07.08.2016.
 */
public class MainPagePlainConverter extends RowConverter implements Converter<MainPlainPage, Document> {

    public MainPagePlainConverter(Logger logger) {
        super(logger);
    }

    @Override
    public MainPlainPage convert(Document input) {
        List<Row> rows = Stream.of(input.select(Selectors.ROWS))
                .flatMap(elements -> elements.stream())
                .skip(1)
                .map(element -> toRow(element))
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
        MainPlainPage page = new MainPlainPage(rows);

        return page;
    }

    private Row toRow(Element element) {
        Elements tds = element.getElementsByTag(Selectors.TD);
        Row output = parseRow(tds);

        return output;
    }

    /* =================== */

    public interface Selectors extends Serializable {

        String ROWS = "#index tr";
        String TD = "td";

    }

}

File: api/src/main/kotlin/org/team619/rutor/converter/MainPagePlainConverter.kt
package org.team619.rutor.converter

import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.team619.rutor.core.Converter
import org.team619.rutor.core.Logger
import org.team619.rutor.model.MainPlainPage
import org.team619.rutor.model.Row
import org.team619.rutor.util.Constants.Companion.PLAIN_ROWS
import org.team619.rutor.util.Constants.Companion.TD
import java.util.*

/**
 * Created by BORIS on 07.08.2016.
 */
class MainPagePlainConverter(logger: Logger) : RowConverter(logger), Converter<MainPlainPage, Document> {

    override fun convert(input: Document): MainPlainPage {
        val rows = input.select(PLAIN_ROWS).drop(1)
                .map { element -> toRow(element) }
                .filter({ Objects.nonNull(it) })
                .map { it!! }

        return MainPlainPage(rows)
    }

    private fun toRow(element: Element): Row? {
        return parseRow(
                element.getElementsByTag(TD))
    }

    //274289764 = Plain page text
    override fun support(element: Element): Boolean {
        val text = element.select("#index > h2").last().text().hashCode()
        val size = element.select("#index > table > tbody > tr").size

        return 274289764 == text && size >= 10
    }

}
